linker: move IDT_LIST region

JordanYates opened the pull request:
Move the IDT_LIST memory region to the location recommended by
`intlist.ld`. The documentation specifies that this region should not
overlap other regions, and there is no guarantee that the area after the
`SRAM` region is not used. The end of the address space is much less
likely to be a valid RAM address.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
linker: arm: move IDT_LIST region

Move the IDT_LIST memory region to the location recommended by
`intlist.ld`. The documentation specifies that this region should not
overlap other regions, and there is no guarantee that the area after the
`SRAM` region is not used. The end of the address space is much less
likely to be a valid RAM address.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

stephanosio reviewed (commented):
Although this is a very ugly hack, it seems to be the best we can do for now unless we rework the whole bogus IDT mechanism (which, in itself, is a sort of hack).

While we are at it, these should be updated too:
https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/soc/xtensa/intel_s1000/memory.h#L66

https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/soc/xtensa/intel_adsp/cavs_v18/include/soc/memory.h#L87

https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/soc/xtensa/intel_adsp/cavs_v15/include/soc/memory.h#L90

https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/soc/xtensa/intel_adsp/cavs_v20/include/soc/memory.h#L87

https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/soc/xtensa/intel_adsp/cavs_v25/include/soc/memory.h#L87

yashi commented:
Just curious.  ARM has so-called high vector address (ref SCTLR.V) and many board map memory at 0xFFFF0000.  Does it conflict with the address `0xFFFFF7FF` which this commit uses?

https://developer.arm.com/documentation/ddi0406/b/System-Level-Architecture/The-System-Level-Programmers--Model/Exceptions/Exception-vectors-and-the-exception-base-address?lang=en

JordanYates commented:
> Just curious. ARM has so-called high vector address (ref SCTLR.V) and many boards map memory at 0xFFFF0000. Does it conflict with the address 0xFFFFF7FF which this commit uses?

I wasn't aware of this feature, but it starts at 0xFFFF0000, and it seems unlikely that any SoC would have 63487 bytes of vector table to end up in this region.

Can you link one of the boards that maps memory to 0xFFFF000? Does the memory run all the way to 0xFFFFFFFF?

stephanosio commented:
> Just curious. ARM has so-called high vector address (ref SCTLR.V) and many boards map memory at 0xFFFF0000. Does it conflict with the address `0xFFFFF7FF` which this commit uses?

Even if `HIVECS` is enabled, the vector table itself is only 32B long, so there should be no overlap.

Of course, if an SoC that supports `HIVECS` has a memory region defined spanning to the end of the 32-bit memory space, that would be problematic; but, in reality, that is quite unlikely.

For Cortex-R devices, `HIVECS` will almost never be used (it is not recommended by Arm either).
For Cortex-A devices, I have not seen any devices that are designed to use `HIVECS`. Most Cortex-A SoCs map high memory regions to DDR, so if your DDR memory is large enough, you might reach the end of address space; but that is a rather special case.

yashi commented:
hmmm... I don't remember any off the top of my head.  I thought many old boards used to use it and one of OMAP or XSCALE CPU did but no link at hands. (OK I admit it was long time ago before Cortex).

Quick greps on both Linux and U-Boot with `0xffff0000`, `VECTOR_BASE`, or `high vector` shows some code but not exactly which.

Apologize for the last comment without any concrete example.

Commit:
linker: xtensa: move IDT_LIST region

Move the IDT_LIST memory region to the location recommended by
`intlist.ld`. The documentation specifies that this region should not
overlap other regions, and there is no guarantee that the area after the
`SRAM` region is not used. The end of the address space is much less
likely to be a valid RAM address.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

povergoing commented:
I not sure, but could it be made configurable? like at Kconfig/DTS

stephanosio commented:
> I not sure, but could it be made configurable? like at Kconfig/DTS

That sounds quite reasonable; but, noting that this PR is merely synchronising the handling of `IDT_LIST` across multiple archs (i.e. following the existing convention used by other archs), it may be beyond the scope of this PR.

dcpleung commented:
INT_LIST is only used for intermediate linking steps. It is set at beyond RAM to avoid being linked within SRAM addresses so it won't move the symbols around (as they are discarded at final linking). Just curious if this is causing any issue.
